packages/base-controller/src/next/BaseController.ts

```diff
@@ -164,19 +171,37 @@ export type ControllerEvents<
 export class BaseController<
   ControllerName extends string,
   ControllerState extends StateConstraint,
-  // TODO: Either fix this lint violation or explain why it's necessary to ignore.
-  // eslint-disable-next-line @typescript-eslint/naming-convention
-  messenger extends RestrictedMessenger<
+  ControllerMessenger extends Messenger<
     ControllerName,
-    ActionConstraint | ControllerActions<ControllerName, ControllerState>,
-    EventConstraint | ControllerEvents<ControllerName, ControllerState>,
-    string,
-    string
+    ActionConstraint,
+    EventConstraint,
+    // Use `any` to allow any parent to be set. `any` is harmless in a type constraint anyway,
+    // it's the one totally safe place to use it.
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    any
   >,
 > {
+  /**
+   * The controller state.
+   */
   #internalState: ControllerState;
 
-  protected messagingSystem: messenger;
+  /**
+   * The controller messenger. This is used to interact with other parts of the application.
+   */
+  protected messagingSystem: ControllerMessenger;
+
+  /**
+   * The controller messenger.
+   *
+   * This is the same as the `messagingSystem` property, but has a type that only lets us use
+   * actions and events that are part of the `BaseController` class.
+   */
+  readonly #messenger: Messenger<
```

Comment:
I added a duplicate reference to the messenger here as a private instance variable just to avoid the need to cast it each time it's internally referenced.